### PR TITLE
45746 Send B2C token to FSS for daily tracings when logged in

### PR DESCRIPTION
**Pull request by Ashish15251**

### CODE REVIEW

**Review by usha13529 — CHANGES_REQUESTED** [review]
Please fix the review comments.

File: UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Web/Services/NMDataService.cs [hunk 1]
```diff
@@ -126,11 +127,20 @@ public async Task<List<ShowDailyFilesResponseModel>> GetDailyBatchDetailsFiles(s
         {
             try
             {
-                string accessToken = await _authFssTokenProvider.GenerateADAccessToken(correlationId);
+                string accessToken = await _authFssTokenProvider.GenerateADAccessToken(_userService.IsDistributorUser, correlationId);
+
+                _logger.LogInformation(EventIds.ShowDailyFilesResponseStarted.ToEventId(), "Maritime safety information request to get daily NM files response started for daily user:{_userService.SignInName} and Identity:{_userService.UserIdentifier} with _X-Correlation-ID:{correlationId}", _userService.SignInName, _userService.UserIdentifier, correlationId);
```
[the remark this answers is not in the document]
Comment (Ashish15251) [reply UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Web/Services/NMDataService.cs:132]: Changes done

File: UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Common.UnitTests/Helpers/NMHelperTest.cs [hunk 2]
```diff
@@ -16,9 +16,9 @@ namespace UKHO.MaritimeSafetyInformation.Common.UnitTests.Helpers
     public class NMHelperTest
     {
         [Test]
-        public void WhenNMHelperCallsListFilesResponse_ThenConversionIsCorrect()
+        public void WhenNMHelperCallsListFilesResponseForPublicUser_ThenConversionIsCorrect()
         {
-            BatchSearchResponse searchResult = SetSearchResultForWeekly();
+            BatchSearchResponse searchResult = SetSearchResultForWeeklyForPublicUser();
```
[the remark this answers is not in the document]
Comment (Ashish15251) [reply UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Common.UnitTests/Helpers/NMHelperTest.cs:22]: Changes Done

File: UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Web/Services/NMDataService.cs [hunk 3]
```diff
@@ -126,11 +127,20 @@ public async Task<List<ShowDailyFilesResponseModel>> GetDailyBatchDetailsFiles(s
...
 
-                const string searchText = $" and $batch(Frequency) eq 'daily'";
+                if (_userService.IsDistributorUser)
+                {
+                    searchText = $" and $batch(Frequency) eq 'Daily' and $batch(Content) eq 'Tracings' ";
+                }
+                else
+                {
+                    searchText = $" and $batch(Frequency) eq 'Daily' and $batch(Content) eq null ";
+                }
```
[the remark this answers is not in the document]
Comment (Ashish15251) [reply UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Web/Services/NMDataService.cs:143]: Changes Done

File: UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Web/Views/NoticestoMariners/ShowWeeklyFilesList.cshtml [hunk 4]
```diff
@@ -1,8 +1,45 @@
...
 @{
-    var a = Model;
+    var publicRole = Model.Where(x => x.IsDistributorUser.Equals(false));
 }
+
+@if (Model != null && Model.Count > 0)
+{
+    @if (ViewBag.IsDistributor)
+    {
+        <caption class="table-caption">Partner</caption>
```
[the remark this answers is not in the document]
Comment (Ashish15251) [reply UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Web/Views/NoticestoMariners/ShowWeeklyFilesList.cshtml:11]: Changes Done

File: UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Web/Controllers/NoticestoMarinersController.cs [hunk 5]
```diff
@@ -97,7 +99,7 @@ public async Task<IActionResult> ShowDailyFilesAsync()
             {
                 _logger.LogInformation(EventIds.ShowDailyFilesRequest.ToEventId(), "Maritime safety information request to show daily NM files started for _X-Correlation-ID:{correlationId}", GetCurrentCorrelationId());
 
-                ViewBag.IsDistributor = _userService.IsDistributorUser;
+                ViewBag.IsDistributor = _userService.IsDistributorUser;//true - user login (create new user token), false - existing working.
```
[the remark this answers is not in the document]
Comment (Ashish15251) [reply UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Web/Controllers/NoticestoMarinersController.cs:102]: Changes Done

File: UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Web/Services/UserService.cs [hunk 6]
```diff
@@ -1,23 +1,44 @@
...
+            get
+            {
                 return _httpContextAccessor.HttpContext.User.Identity.IsAuthenticated
                     && _httpContextAccessor.HttpContext.User.IsInRole(DistributorRoleName);
             }
-        }     
+        }
+
+        public string UserIdentifier
+        {
```
[the remark this answers is not in the document]
Comment (Ashish15251) [reply UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Web/Services/UserService.cs:29]: Changes Done

File: UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Web/Services/UserService.cs [hunk 7]
```diff
@@ -1,23 +1,44 @@
...
+            get
+            {
+                return Convert.ToString(_httpContextAccessor.HttpContext.User.FindFirstValue("http://schemas.microsoft.com/identity/claims/objectidentifier"));
+            }
+        }
+
+        public string SignInName
+        {
```
[the remark this answers is not in the document]
Comment (Ashish15251) [reply UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Web/Services/UserService.cs:37]: Changes Done

File: Tests/UKHO.MaritimeSafetyInformation.Web.UnitTests/Services/UserServiceTests.cs [hunk 8]
```diff
@@ -52,7 +52,7 @@ public void WhenUserHasRole_ThenIsDistributorReturnsTrue()
             var user = new ClaimsPrincipal(new ClaimsIdentity(new Claim[]
                 {
                     new Claim(ClaimTypes.Name, "Distributor"),
-                   new Claim(ClaimTypes.Role, DistributorRoleName)
+                   new Claim(ClaimTypes.Role, DistributorRoleName),
```
Comment (usha13529) [inline Tests/UKHO.MaritimeSafetyInformation.Web.UnitTests/Services/UserServiceTests.cs:55]:
```suggestion
                   new Claim(ClaimTypes.Role, DistributorRoleName)
```

**Review by usha13529 — CHANGES_REQUESTED** [review]
Few more comments to fix please.

File: UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Web/Services/NMDataService.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (Ashish15251) [reply UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Web/Services/NMDataService.cs]: changes done

File: UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Common/Helpers/NMHelper.cs [hunk 9]
```diff
@@ -10,12 +10,24 @@ public static class NMHelper
...
+                BatchDetails distBatch = distributorBatchDetail.OrderByDescending(t => t.BatchPublishedDate).FirstOrDefault();
+                BatchDetails publicBatch = publicBatchDetail.OrderByDescending(t => t.BatchPublishedDate).FirstOrDefault();
+                if (distBatch != null)
+                {
+                    batchDetailsList.Add(distBatch);
+                }
+
+                if (publicBatch != null)
+                    batchDetailsList.Add(publicBatch);
+
```
[the remark this answers is not in the document]
Comment (Ashish15251) [reply UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Common/Helpers/NMHelper.cs]: changes done

File: UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Web/Services/NMDataService.cs [hunk 10]
```diff
@@ -140,30 +145,30 @@ public async Task<List<ShowDailyFilesResponseModel>> GetDailyBatchDetailsFiles(s
 
                 if (searchResult != null && searchResult.Entries != null && searchResult.Entries.Count > 0)
                 {
-                    _logger.LogInformation(EventIds.ShowDailyFilesResponseDataFound.ToEventId(), "Maritime safety information request to get daily NM files response data found for _X-Correlation-ID:{correlationId}", correlationId);
+                    _logger.LogInformation(EventIds.ShowDailyFilesResponseDataFound.ToEventId(), "Maritime safety information request to get daily NM files response started for daily user:{SignInName} and Identity:{userId} with _X-Correlation-ID:{correlationId}", _userService.SignInName ?? "Public", _userService.UserIdentifier, correlationId);
```
[the remark this answers is not in the document]
Comment (Ashish15251) [reply UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Web/Services/NMDataService.cs]: changes done

File: UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Web/Services/NMDataService.cs [hunk 11]
```diff
@@ -140,30 +145,30 @@ public async Task<List<ShowDailyFilesResponseModel>> GetDailyBatchDetailsFiles(s
...
                     List<ShowDailyFilesResponseModel> showDailyFilesResponses = NMHelper.GetDailyShowFilesResponse(searchResult);
                     return showDailyFilesResponses;
                 }
                 else
                 {
-                    _logger.LogError(EventIds.ShowDailyFilesResponseDataNotFound.ToEventId(), "Maritime safety information request to get daily NM files response data not found for _X-Correlation-ID:{correlationId}", correlationId);
+                    _logger.LogError(EventIds.ShowDailyFilesResponseDataNotFound.ToEventId(), "Maritime safety information request to get daily NM files response data not found for daily user:{SignInName} and Identity:{userId} with _X-Correlation-ID:{correlationId}", _userService.SignInName ?? "Public", _userService.UserIdentifier, correlationId);
```
[the remark this answers is not in the document]
Comment (Ashish15251) [reply UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Web/Services/NMDataService.cs]: changes done

File: UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Web/Services/NMDataService.cs [hunk 12]
```diff
@@ -140,30 +145,30 @@ public async Task<List<ShowDailyFilesResponseModel>> GetDailyBatchDetailsFiles(s
...
                     throw new InvalidDataException("Invalid data received for daily NM files");
                 }
 
 
             }
             catch (Exception ex)
             {
-                _logger.LogError(EventIds.ShowDailyFilesResponseFailed.ToEventId(), "Maritime safety information request to get daily NM files failed to return data with exception:{exceptionMessage} for _X-Correlation-ID:{CorrelationId}", ex.Message, correlationId);
-                throw;
+                _logger.LogError(EventIds.ShowDailyFilesResponseFailed.ToEventId(), "Maritime safety information request to get daily NM files failed to return data with exception:{exceptionMessage} for daily user:{SignInName} and Identity:{userId} with _X-Correlation-ID:{correlationId}", ex.Message, _userService.SignInName ?? "Public", _userService.UserIdentifier, correlationId);
```
[the remark this answers is not in the document]
Comment (Ashish15251) [reply UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Web/Services/NMDataService.cs]: changes done

File: Tests/UKHO.MaritimeSafetyInformation.Web.UnitTests/Services/UserServiceTests.cs [hunk 13]
```diff
@@ -67,5 +67,52 @@ public void WhenUserHasRole_ThenIsDistributorReturnsTrue()
 
             Assert.AreEqual(true, userService.IsDistributorUser);
         }
+
+        [Test]
+        public void WhenUserIsauthenticated_ThenGetSignInNameAndUserIdentityReturnsTrue()
```
[the remark this answers is not in the document]
Comment (Ashish15251) [reply Tests/UKHO.MaritimeSafetyInformation.Web.UnitTests/Services/UserServiceTests.cs]: changes done

File: Tests/UKHO.MaritimeSafetyInformation.Web.UnitTests/Services/UserServiceTests.cs (hunk 8 above)
Comment (Ashish15251) [reply Tests/UKHO.MaritimeSafetyInformation.Web.UnitTests/Services/UserServiceTests.cs]: changes done

File: Tests/UKHO.MaritimeSafetyInformation.Web.UnitTests/Services/UserServiceTests.cs [hunk 14]
```diff
@@ -67,5 +67,52 @@ public void WhenUserHasRole_ThenIsDistributorReturnsTrue()
...
+            UserService userService = new(mockHttpContextAccessor);
+
+            Assert.AreEqual("TestUser", userService.SignInName);
+            Assert.AreEqual("f457520b-0f1c-44cc-9b5c-2113e2ba1234", userService.UserIdentifier);
+
+        }
+
+        [Test]
+        public void WhenUserIsUnauthenticated_ThenGetSignInNameAndUserIdentityReturnsFalse()
+        {
```
[the remark this answers is not in the document]
Comment (Ashish15251) [reply Tests/UKHO.MaritimeSafetyInformation.Web.UnitTests/Services/UserServiceTests.cs]: changes done

File: Tests/AutoTests/FunctionalTest/functional-weeklyfiledownload.spec.ts [hunk 15]
```diff
@@ -22,7 +25,7 @@ test.describe("Goto maritime-safety-information Notice To Mariner Page to Check
     var fileName = name[0];
 
     const newPageUrl = await (await page.$("[id^='download'] > a")).getAttribute('href');
-    expect(newPageUrl).toContain(`NoticesToMariners/DownloadFile?fileName=${fileName}`);
+    expect(newPageUrl).toContain(`NoticesToMariners/DownloadWeeklyFile?fileName=${fileName}`);
```
[the remark this answers is not in the document]
Comment (Pradeep15558) [reply Tests/AutoTests/FunctionalTest/functional-weeklyfiledownload.spec.ts]: Resolved

File: Tests/AutoTests/FunctionalTest/functional-RadioNavigationalWarningsUpdateRNW.spec.ts [hunk 16]
```diff
@@ -20,6 +20,7 @@ test.describe("Create new radio navigational warnings record", ()=> {
     await radioNavigationalWarnings.clearInput(radioNavigationalWarnings.reference);
     await radioNavigationalWarnings.clearInput(radioNavigationalWarnings.content);
     await radioNavigationalWarnings.editRNW();
+    await page.waitForTimeout(5000);
```
Comment (nevillejrbrown) [inline Tests/AutoTests/FunctionalTest/functional-RadioNavigationalWarningsUpdateRNW.spec.ts:23]: Please don't use waitForTimeout().  Playwright docs state it should only be used for debugging (as per https://playwright.dev/docs/api/class-page#page-wait-for-timeout)

File: Tests/AutoTests/FunctionalTest/functional-weeklyfiledownload.spec.ts [hunk 17]
```diff
@@ -48,9 +51,12 @@ test.describe("Goto maritime-safety-information Notice To Mariner Page to Check
     await notice.checkurl(notice.menuValueAddedResellers,'NoticesToMariners/Resellers','Notices to Mariners - Value Added Resellers')
     await notice.checkurl(notice.menuAbout,'NoticesToMariners/About','About Notices to Mariners')
     await notice.checkurl(notice.radioNavigationalWarnings,'RadioNavigationalWarnings','Radio Navigational Warnings')
+    await page.waitForTimeout(5000);
```
Comment (nevillejrbrown) [inline Tests/AutoTests/FunctionalTest/functional-weeklyfiledownload.spec.ts:54]: Don't use waitForTimeout(), as above

File: Tests/AutoTests/FunctionalTest/functional-weeklyfiledownload.spec.ts [hunk 18]
```diff
@@ -61,4 +67,15 @@ test.describe("Goto maritime-safety-information Notice To Mariner Page to Check
   await noticeFileDownload.verifyleisureFoliosFileNameDownload();
   
 })
+
+test('Should Goto Notices to Mariner Page for Daily download File with Distributor Login', async ({ page, context }) => {
```
Comment (nevillejrbrown) [inline Tests/AutoTests/FunctionalTest/functional-weeklyfiledownload.spec.ts:71]: It's confusing that these daily download tests are in a file called "weeklyfiledownload".  Can they be moved?

**Review by nevillejrbrown — CHANGES_REQUESTED** [review]
Please seem comments

File: UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Web/Controllers/NoticestoMarinersController.cs [hunk 19]
```diff
@@ -1,8 +1,10 @@
 ﻿using Microsoft.AspNetCore.Mvc;
+using Microsoft.Identity.Web;
 using UKHO.MaritimeSafetyInformation.Common.Helpers;
 using UKHO.MaritimeSafetyInformation.Common.Logging;
 using UKHO.MaritimeSafetyInformation.Common.Models.NoticesToMariners;
 using UKHO.MaritimeSafetyInformation.Web.Services.Interfaces;
+using Microsoft.Identity.Web;
```
[the remark this answers is not in the document]
Comment (Ashish15251) [reply UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Web/Controllers/NoticestoMarinersController.cs]: changes done

File: Tests/UKHO.MaritimeSafetyInformation.Web.UnitTests/Controllers/NoticestoMarinersControllerTest.cs [hunk 20]
```diff
@@ -29,6 +31,7 @@ public void Setup()
             _fakeContextAccessor = A.Fake<IHttpContextAccessor>();
             _fakeNMDataService = A.Fake<INMDataService>();
             _fakeUserService = A.Fake<IUserService>();
+            _fakeTokenAcquisition = A.Fake<ITokenAcquisition>();
```
[the remark this answers is not in the document]
Comment (Ashish15251) [reply Tests/UKHO.MaritimeSafetyInformation.Web.UnitTests/Controllers/NoticestoMarinersControllerTest.cs]: changes done

File: Tests/AutoTests/FunctionalTest/functional-RadioNavigationalWarningsUpdateRNW.spec.ts (hunk 16 above)
Comment (Pradeep15558) [reply Tests/AutoTests/FunctionalTest/functional-RadioNavigationalWarningsUpdateRNW.spec.ts:23]: Resolved

File: Tests/AutoTests/FunctionalTest/functional-weeklyfiledownload.spec.ts (hunk 17 above)
Comment (Pradeep15558) [reply Tests/AutoTests/FunctionalTest/functional-weeklyfiledownload.spec.ts]: Resolved

File: Tests/AutoTests/FunctionalTest/functional-weeklyfiledownload.spec.ts (hunk 18 above)
Comment (Pradeep15558) [reply Tests/AutoTests/FunctionalTest/functional-weeklyfiledownload.spec.ts]: Resolved

**Review by usha13529 — APPROVED** [review]
Thanks for the changes @Ashish15251. Looks good to me.

Need to verify FT's in pipeline once FSS QA has latest changes required for tracings data.

File: UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Web/Services/NMDataService.cs [hunk 21]
```diff
@@ -190,11 +192,15 @@ public async Task<List<ShowDailyFilesResponseModel>> GetDailyBatchDetailsFiles(s
...
+                string accessToken = await _authFssTokenProvider.GenerateADAccessToken(_userService.IsDistributorUser, correlationId);
 
-                _logger.LogInformation(EventIds.ShowDailyFilesResponseStarted.ToEventId(), "Maritime safety information request to get daily NM files response started with _X-Correlation-ID:{correlationId}", correlationId);
+                _logger.LogInformation(EventIds.ShowDailyFilesResponseStarted.ToEventId(), "Maritime safety information request to get daily NM files response started for daily user:{SignInName} and Identity:{userId} with _X-Correlation-ID:{correlationId}", _userService.SignInName ?? "Public", _userService.UserIdentifier, correlationId);
 
-                const string searchText = $" and $batch(Frequency) eq 'daily'";
+                string searchText = $" and $batch(Frequency) eq 'Daily' and $batch(Content) eq null ";
+                if (_userService.IsDistributorUser)
+                {
+                    searchText = $" and $batch(Frequency) eq 'Daily' and $batch(Content) eq 'Tracings' ";
```
Comment (barrie-cooper) [inline UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Web/Services/NMDataService.cs:202]: There is no unit test that covers this path.  Tests need to be extended to check that the right search text is being provided to `IFileShareService.FSSBatchSearchAsync` please.

**Review by barrie-cooper — COMMENTED** [review]
Just one point about unit tests.

File: UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Web/Services/NMDataService.cs (hunk 21 above)
Comment (usha13529) [reply UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Web/Services/NMDataService.cs:202]: Thanks for highlighting. Added unit test to cover search query path for distributor.

**Review by barrie-cooper — APPROVED** [review]
LGTM, thanks

**Pull request merged**